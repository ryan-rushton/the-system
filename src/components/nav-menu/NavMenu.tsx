import { faBars } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { CSSProperties, RefObject, useCallback, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { pointsOfInterest } from '../../PointsOfInterest';
import { SystemContext } from '../../context/SystemContext';
import { useClickOutside } from '../../hooks/useClickOutside';
import { getDistanceToTop } from '../../utils/DomUtil';
import styles from './NavMenu.module.scss';
import { NavMenuSubsection } from './NavMenuSubsection';
import { InfoMenu } from './sub-menus/InfoMenu';
import { PointOfInterestButton } from './sub-menus/PointOfInterestButton';

type OpenSubsectionState = 'info' | 'nav' | undefined;

/**
 * A sliding nav menu. When the burger icon is clicked the menu will slide into view. Each section will slide
 * open on clicking the title for it, any other open section will be closed at this point.
 *
 * There are no unit tests for this unfortunately as all the logic is based around extracting details from refs.
 */
export function NavMenu({
  orbitsVisible,
  followedPointOfInterest,
  onChangeSystemSize,
  onOrbitsVisibleChange,
  onFollowPointOfInterest,
}: {
  /** Whether the red orbit lines are visible. */
  orbitsVisible: boolean;
  /** The point of interest currently being followed. */
  followedPointOfInterest?: { ref: RefObject<HTMLDivElement> };
  /** Change handler for whether the red orbit lines are visible. */
  onOrbitsVisibleChange(orbitsVisible: boolean): void;
  /** Change handler for normalising km per pixel. */
  onChangeSystemSize(systemSizeContext: SystemContext): void;
  /** A handler for following a point of interest around the solar system. */
  onFollowPointOfInterest(pointOfInterest: { ref: RefObject<HTMLDivElement> }): void;
}) {
  const [menuVisible, setMenuVisible] = useState(false);
  const [openSubsection, setOpenSubsection] = useState<OpenSubsectionState>();
  const containerRef = useRef<HTMLDivElement>(null);
  const menuRef = useRef<HTMLDivElement>(null);
  const buttonRef = useRef<HTMLButtonElement>(null);
  const { t } = useTranslation();

  const onMenuClick = useCallback(() => setMenuVisible((oldState) => !oldState), []);
  const onSubsectionClick = (clicked: 'info' | 'nav') =>
    setOpenSubsection((oldState) => (clicked === oldState ? undefined : clicked));
  useClickOutside(containerRef, menuVisible, () => setMenuVisible(false));

  const leftEdgeOfContainerToRightViewport =
    window.innerWidth - (containerRef.current?.getBoundingClientRect().left || 0);
  const rightEdgeOfButtonToRightViewport = window.innerWidth - (buttonRef.current?.getBoundingClientRect().right || 0);
  // on first render we dont have the offset so just push it way off the screen
  const transformDistance = menuRef.current
    ? `calc(-${menuRef.current.getBoundingClientRect().width}px + ${rightEdgeOfButtonToRightViewport}px + 4px)`
    : '-100vw';

  const transformStyles: CSSProperties = {
    maxHeight: menuRef.current ? `calc(100vh - ${getDistanceToTop(menuRef.current)}px)` : undefined,
    // the 2vw is because we have a 2vw right indent on the navMenu style
    transform: menuVisible ? `translateX(${transformDistance})` : `translateX(${leftEdgeOfContainerToRightViewport}px)`,
  };

  return (
    <div ref={containerRef} data-testid="nav-menu-container" className={styles.nav}>
      <div className={styles.header}>
        <button ref={buttonRef} data-testid="nav-menu-button" className={styles.headerButton} onClick={onMenuClick}>
          <FontAwesomeIcon icon={faBars} />
        </button>
      </div>
      <div data-testid="nav-menu" className={styles.menu} ref={menuRef} style={transformStyles}>
        <NavMenuSubsection
          title={t('info')}
          isVisible={openSubsection === 'info'}
          canTabInto={menuVisible}
          onHeaderClick={() => onSubsectionClick('info')}
        >
          <InfoMenu
            isVisible={menuVisible && openSubsection === 'info'}
            orbitsVisible={orbitsVisible}
            onChangeSystemSize={onChangeSystemSize}
            onOrbitsVisibleChange={onOrbitsVisibleChange}
          />
        </NavMenuSubsection>
        <NavMenuSubsection
          title={t('navigation')}
          isVisible={openSubsection === 'nav'}
          canTabInto={menuVisible}
          onHeaderClick={() => onSubsectionClick('nav')}
        >
          <div className={styles.navButtons}>
            {Object.values(pointsOfInterest).map((poi) => (
              <PointOfInterestButton
                isBeingFollowed={poi === followedPointOfInterest}
                isVisible={menuVisible && openSubsection === 'nav'}
                key={poi.id}
                pointOfInterest={poi}
                onPoiClick={() => onFollowPointOfInterest(poi)}
              />
            ))}
          </div>
        </NavMenuSubsection>
      </div>
    </div>
  );
}
