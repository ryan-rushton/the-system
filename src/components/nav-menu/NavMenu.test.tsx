import { render, screen } from '@testing-library/react';
import React from 'react';
import NavMenu from './NavMenu';

describe('NavMenu', () => {
  test('it renders', () => {
    render(
      <NavMenu
        orbitsVisible={false}
        onChangeSystemSize={jest.fn()}
        onFollowPointOfInterest={jest.fn()}
        onOrbitsVisibleChange={jest.fn()}
      />,
    );

    expect(screen.getByTestId('nav-menu-button')).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    render(
      <NavMenu
        orbitsVisible={false}
        onChangeSystemSize={jest.fn()}
        onFollowPointOfInterest={jest.fn()}
        onOrbitsVisibleChange={jest.fn()}
      />,
    );

    expect(screen.getByTestId('nav-menu-container')).toMatchSnapshot();
  });
});
