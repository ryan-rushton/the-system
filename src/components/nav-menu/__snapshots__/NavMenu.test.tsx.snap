// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu it matches snapshot 1`] = `
<div
  class="nav"
  data-testid="nav-menu-container"
>
  <div
    class="header"
  >
    <button
      class="headerButton"
      data-testid="nav-menu-button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
  <div
    class="menu"
    data-testid="nav-menu"
    style="transform: translateX(100vw);"
  >
    <div
      class="subsection"
    >
      <button
        class="title"
      >
        Info
      </button>
      <div
        class="content"
        style="max-height: 0;"
      >
        <div
          data-testid="info-menu"
        >
          <div
            class="heading"
          >
            Time
          </div>
          <div
            class="stat"
            data-testid="time"
          >
            1 s = 1 day
          </div>
          <div
            class="heading"
          >
            Distance Between Planets
          </div>
          <div
            class="stat"
            data-testid="planet-distance"
          >
            1 pixel = 100,000 km
          </div>
          <div
            class="heading"
          >
            Distance Between
Planets And Moons
          </div>
          <div
            class="stat"
            data-testid="satellite-distance"
          >
            1 pixel = 10,000 km
          </div>
          <div
            class="heading"
          >
            Planet Size
          </div>
          <div
            class="stat"
            data-testid="planet-size"
          >
            1 pixel = 2,000 km
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              aria-label="Normalise Distance"
              class="button"
            >
              Normalise Distance
            </button>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              aria-label="Show Orbits"
              class="button"
            >
              Show Orbits
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="subsection"
    >
      <button
        class="title"
      >
        Navigation
      </button>
      <div
        class="content"
        style="max-height: 0;"
      >
        <div
          class="navButtons"
        >
          <button
            aria-label="Sun"
            class="navItem"
            data-testid="sun-button"
          >
            <span>
              Sun
            </span>
          </button>
          <button
            aria-label="Mercury"
            class="navItem"
            data-testid="mercury-button"
          >
            <span>
              Mercury
            </span>
          </button>
          <button
            aria-label="Venus"
            class="navItem"
            data-testid="venus-button"
          >
            <span>
              Venus
            </span>
          </button>
          <button
            aria-label="Earth"
            class="navItem"
            data-testid="earth-button"
          >
            <span>
              Earth
            </span>
          </button>
          <button
            aria-label="Mars"
            class="navItem"
            data-testid="mars-button"
          >
            <span>
              Mars
            </span>
          </button>
          <button
            aria-label="The Belt"
            class="navItem"
            data-testid="theBelt-button"
          >
            <span>
              The Belt
            </span>
          </button>
          <button
            aria-label="Jupiter"
            class="navItem"
            data-testid="jupiter-button"
          >
            <span>
              Jupiter
            </span>
          </button>
          <button
            aria-label="Saturn"
            class="navItem"
            data-testid="saturn-button"
          >
            <span>
              Saturn
            </span>
          </button>
          <button
            aria-label="Uranus"
            class="navItem"
            data-testid="uranus-button"
          >
            <span>
              Uranus
            </span>
          </button>
          <button
            aria-label="Neptune"
            class="navItem"
            data-testid="neptune-button"
          >
            <span>
              Neptune
            </span>
          </button>
          <button
            aria-label="Pluto"
            class="navItem"
            data-testid="pluto-button"
          >
            <span>
              Pluto
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
